libs/template/class/src/lib/class.directive.ts

```diff
@@ -0,0 +1,197 @@
+import {
+  Directive,
+  ElementRef,
+  Input,
+  IterableChanges,
+  IterableDiffer,
+  IterableDiffers,
+  KeyValueChanges,
+  KeyValueDiffer,
+  KeyValueDiffers,
+  OnDestroy,
+  OnInit,
+  Output,
+  Renderer2,
+  ɵisListLikeIterable as isListLikeIterable
+} from '@angular/core';
+import {
+  accumulateObservables,
+  RxStrategyNames,
+  RxStrategyProvider
+} from '@rx-angular/cdk';
+import { coerceAllFactory } from '@rx-angular/cdk/coercing';
+import {
+  BehaviorSubject,
+  isObservable,
+  Observable,
+  ObservableInput,
+  ReplaySubject,
+  Subject,
+  Subscription
+} from 'rxjs';
+import { map, switchAll, switchMap, withLatestFrom } from 'rxjs/operators';
+
+type NgClassValues =
+  | string[]
+  | Set<string>
+  | Record<string, boolean>
+  | string
+  | null
+  | undefined;
+
+type RxClassInput =
+  | Observable<NgClassValues>
+  | Record<string, Observable<boolean>>;
+
+@Directive({ selector: '[rxClass]' })
+export class ClassDirective implements OnInit, OnDestroy {
+  private iterableDiffer: IterableDiffer<string> | null = null;
+  private keyValueDiffer: KeyValueDiffer<string, any> | null = null;
+  private rawClass: NgClassValues = null;
+  private newInput = false;
+  private sub = new Subscription();
+  private classValueChanged = false;
+
+  constructor(
+    private readonly iterableDiffers: IterableDiffers,
+    private readonly keyValueDiffers: KeyValueDiffers,
+    private readonly ngEl: ElementRef,
+    private readonly renderer: Renderer2,
+    private readonly strategyProvider: RxStrategyProvider
+  ) {}
+
+  private rxClass = coerceAllFactory<NgClassValues>(
+    () => new ReplaySubject(1),
+    switchAll()
+  );
+
+  @Input('rxClass')
+  set rxClass$(value$: RxClassInput) {
+    this.newInput = true;
+    this.rxClass.next(
+      (isObservable(value$) ? value$ : accumulateObservables(value$)).pipe(
+        map((value) => (typeof value === 'string' ? [value] : value))
+      )
+    );
+  }
+
+  private strategy = coerceAllFactory(
+    () => new BehaviorSubject<ObservableInput<string> | string>(this.strategyProvider.primaryStrategy),
+    switchAll()
+  );
+
+  @Input('rxClassStrategy')
+  set strategy$(
+    strategy: RxStrategyNames<string> | Observable<RxStrategyNames<string>>
+  ) {
+    this.strategy.next(strategy);
+  }
+
+  @Output() readonly renderCallback = new Subject<void>();
+
+  ngOnInit(): void {
+    this.sub.add(
+      this.rxClass.values$
+        .pipe(
+          withLatestFrom(this.strategy.values$),
+          switchMap(([value, strategy]) =>
+            this.strategyProvider.schedule(
+              () => {
+                this.classValueChanged = false;
+                this.rawClass = value;
+                if (this.newInput) {
```

Comment:
Why do we need these `newInput` + `classValueChanged` variables? reacting to `rxClass` observable events to update DOM values would be enough, am I missing something?

Reply:
The `classValueChanged` is there because of the `renderCallback` `@Output`. The `newInput` property is used to initialize the `differs` using the first value passed at the very beginning. I will rewrite that a bit so that it will be easier to read. The `differs` behaviour is kind of unusual to me as well at some point tbh 😄

Reply:
@hoebbelsB what is your proposal for this thing? I am not quite sure if any change is needed. 🤔

Reply:
I have no proposal, I was just curious what would be the outcome of 'I will rewrite that a bit so that it will be easier to read' :)